nav2_dwb_controller/dwb_controller/src/dwb_controller.cpp

```diff
@@ -13,46 +13,118 @@
 // limitations under the License.
 
 #include "dwb_controller/dwb_controller.hpp"
-#include <string>
+
 #include <chrono>
 #include <memory>
+#include <string>
+
 #include "dwb_core/exceptions.hpp"
 #include "nav_2d_utils/conversions.hpp"
 
 using namespace std::chrono_literals;
-using std::shared_ptr;
 using nav2_tasks::TaskStatus;
-using dwb_core::DWBLocalPlanner;
-using dwb_core::CostmapROSPtr;
-
-#define NO_OP_DELETER [] (auto) {}
 
 namespace nav2_dwb_controller
 {
 
-DwbController::DwbController(rclcpp::executor::Executor & executor)
-: Node("DwbController"),
-  tfBuffer_(get_clock()),
-  tfListener_(tfBuffer_)
+DwbController::DwbController()
+: LifecycleNode("dwb_controller")
 {
-  auto temp_node = std::shared_ptr<rclcpp::Node>(this, [](auto) {});
+  RCLCPP_INFO(get_logger(), "Creating");
 
-  cm_ = std::make_shared<nav2_costmap_2d::Costmap2DROS>("local_costmap", tfBuffer_);
-  executor.add_node(cm_);
-  odom_sub_ = std::make_shared<nav_2d_utils::OdomSubscriber>(*this);
-  vel_pub_ =
-    this->create_publisher<geometry_msgs::msg::Twist>("/cmd_vel", 1);
+  costmap_ros_ = std::make_shared<nav2_costmap_2d::Costmap2DROS>("dwb_controller_local_costmap");
+  costmap_thread_ = std::make_unique<std::thread>(
```

Comment:
The main issue is that Costmap2DORS is currently a ROS node but we're not using it as such. The methods of this contained class are called directly from the users of this class, like DWBLocalPlanner, rather than via ROS mechanisms. So, I see it as part of the implementation and not a top-level node like DWB and the others. In general, I'd prefer that it either be a legit ROS node (in which case it makes sense to me to then compose with the other top-level nodes), or have it be a helper class that takes a ROS node to use for any ROS-related functionality, such as creating subscribers. All that being said, it seemed weird to me to instance the Costmap2DROS in main just to add it to the executor. The alternative was to encapsulate the use of this class and give it its own thread.